### PR TITLE
fix: filename escaping (last character lost sometimes, no hyperlink)

### LINKED ISSUE #162: bug: eza doesn’t show characters in filename after last escaped character
Code extracted from `devtools/dev-create-test-filesystem.sh` and slightly edited:

<details><summary>create files</summary>

```
export TEST_ROOT=test_root
mkdir -p "$TEST_ROOT/file-names"
echo -e "\033[1m[ 3/13]\033[0m Creating file names testcases"

echo -ne "$TEST_ROOT/file-names/ascii: hello" | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/emoji: [🆒]"  | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/utf-8: pâté"  | xargs -0 touch

echo -ne "$TEST_ROOT/file-names/bell: [\a]"         | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/backspace: [\b]"    | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/form-feed: [\f]"    | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/new-line: [\n]"     | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/return: [\r]"       | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/tab: [\t]"          | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/vertical-tab: [\v]" | xargs -0 touch

echo -ne "$TEST_ROOT/file-names/escape: [\033]"              | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/ansi: [\033[34mblue\033[0m]" | xargs -0 touch

echo -ne "$TEST_ROOT/file-names/invalid-utf8-1: [\xFF]"             | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/invalid-utf8-2: [\xc3\x28]"         | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/invalid-utf8-3: [\xe2\x82\x28]"     | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/invalid-utf8-4: [\xf0\x28\x8c\x28]" | xargs -0 touch

echo -ne "$TEST_ROOT/file-names/new-line-dir: [\n]"               | xargs -0 mkdir
echo -ne "$TEST_ROOT/file-names/new-line-dir: [\n]/subfile"       | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/new-line-dir: [\n]/another: [\n]" | xargs -0 touch
echo -ne "$TEST_ROOT/file-names/new-line-dir: [\n]/broken"        | xargs -0 touch
```
</details>

```
$ eza -1 test_root/file-names/
ansi: [\u{1b}[34mblue\u{1b}
ascii: hello
backspace: [\u{8}
bell: [\u{7}
emoji: [🆒]
escape: [\u{1b}
form-feed: [\u{c}
invalid-utf8-1: [�]
invalid-utf8-2: [�(]
invalid-utf8-3: [�(]
invalid-utf8-4: [�(�(]
new-line-dir: [\n
new-line: [\n
return: [\r
tab: [\t
utf-8: pâté
vertical-tab: [\u{b}

$ exa -1 test_root/file-names/
ansi: [\u{1b}[34mblue\u{1b}[0m]
ascii: hello
backspace: [\u{8}]
bell: [\u{7}]
emoji: [🆒]
escape: [\u{1b}]
form-feed: [\u{c}]
invalid-utf8-1: [�]
invalid-utf8-2: [�(]
invalid-utf8-3: [�(]
invalid-utf8-4: [�(�(]
new-line-dir: [\n]
new-line: [\n]
return: [\r]
tab: [\t]
utf-8: pâté
vertical-tab: [\u{b}]

$ ls -1 test_root/file-names/
'ansi: ['$'\033''[34mblue'$'\033''[0m]'
'ascii: hello'
'backspace: ['$'\b'']'
'bell: ['$'\a'']'
'emoji: [🆒]'
'escape: ['$'\033'']'
'form-feed: ['$'\f'']'
'invalid-utf8-1: ['$'\377'']'
'invalid-utf8-2: ['$'\303''(]'
'invalid-utf8-3: ['$'\342\202''(]'
'invalid-utf8-4: ['$'\360''('$'\214''(]'
'new-line: ['$'\n'']'
'new-line-dir: ['$'\n'']'/
'return: ['$'\r'']'
'tab: ['$'\t'']'
'utf-8: pâté'
'vertical-tab: ['$'\v'']'
```

### PR DESCRIPTION
**Pull request by ariasuni**
Follow-up of #17.
Fix #162.

Other bugs fixed/improvements:
- Take control characters outside of ASCII into account (by using [`char::is_control`](https://doc.rust-lang.org/std/primitive.char.html#method.is_control)).
- Hyperlink are now appearing on filenames with escaped characters
- Deduplicated the escaping code by replacing the original code with the optimized one (that I very slightly improved).

### CODE REVIEW

File: src/output/escape.rs [hunk 1]
```diff
@@ -2,25 +2,31 @@ use ansi_term::{ANSIString, Style};
 
 
 pub fn escape(string: String, bits: &mut Vec<ANSIString<'_>>, good: Style, bad: Style) {
-    if string.chars().all(|c| c >= 0x20 as char && c != 0x7f as char) {
+    // if the string has no control character
+    if string.chars().all(|c| !c.is_control()) {
```
Comment (cafkafk) [inline src/output/escape.rs:6]: This is more idiomatic and elegant, nice!

File: src/output/escape.rs [hunk 2]
```diff
@@ -2,25 +2,31 @@ use ansi_term::{ANSIString, Style};
...
-        }
-        else {
-            let s = c.escape_default().collect::<String>();
-            bits.push(bad.paint(s));
+        if c.is_control() {
+            if !regular_char_buff.is_empty() {
+                bits.push(good.paint(std::mem::take(&mut regular_char_buff)));
+            }
+            regular_char_buff.extend(c.escape_default());
+            // biased towards regular characters, we push control characters them immediately
```
Comment (cafkafk) [inline src/output/escape.rs:22]: Slight typo
```suggestion
            // biased towards regular characters, we push control characters immediately
```

File: src/output/file_name.rs [hunk 3]
```diff
@@ -303,59 +303,22 @@ impl<'a, 'dir, C: Colours> FileName<'a, 'dir, C> {
         let file_style = self.style();
         let mut bits = Vec::new();
 
-        self.escape_color_and_hyperlinks(
+        if self.options.embed_hyperlinks == EmbedHyperlinks::On {
+            bits.push(ANSIString::from(format!("\x1B]8;;{}\x1B\x5C", self.file.path.display())));
```
Comment (cafkafk) [inline src/output/file_name.rs:307]: `\x1B]8;;{}\x1B\x5C` is used both here and on line 318, maybe we should introduce a `const` to make the hyperlink format "magic string" easier to understand?

File: src/output/file_name.rs [hunk 4]
```diff
@@ -303,59 +303,22 @@ impl<'a, 'dir, C: Colours> FileName<'a, 'dir, C> {
...
-                    bits.push(good.paint(std::mem::take(&mut buff)));
-                }
-                // biased towards regular characters, so we still collect on first sight of bad char
-                for e in c.escape_default() {
-                    buff.push(e);
-                }
-                bits.push(bad.paint(std::mem::take(&mut buff)));
-            }
-        }
+        bits
```
Comment (cafkafk) [inline src/output/file_name.rs:321]: This is very satisfying, well done!

File: src/output/file_name.rs (hunk 3 above)
Comment (cafkafk) [reply src/output/file_name.rs:307]: Perhaps also do a quick check if it appears in other places in the codebase?

**Review by cafkafk — CHANGES_REQUESTED** [review]
I haven't had time to run this yet, but here are my immediate thoughts.

Also, how did you A/B test this?

File: src/output/file_name.rs (hunk 3 above)
Comment (cafkafk) [reply src/output/file_name.rs:307]: Fixing this could be it's own separate commit if you'd prefer that.

**Comment by ariasuni** [comment]
I added the `file:///` protocol makes the hyperlink feature works in Konsole, but then I realized it’s broken and started investigating [the spec (?)](https://gist.github.com/egmontkob/eb114294efbcd5adb1944c9f3cb5feda).

So after investigation, what needs to be done:

- URL-encode the URL to make it work for «complex» file names
- Add hostname in the URL, e.g. `file://hostname/rest/of/the/path`

I already have working code for the first item that I need to polish and push. The second one should be trivial if I use the correct crate.

**Comment by ariasuni** [comment]
Note: hyperlinks don’t work in Konsole, but same as `ls`, it’s because of a [KDE bug](https://bugs.kde.org/show_bug.cgi?id=461756). It works fine in Kitty.

As said, it works even with weird file names, it fails with non-UTF-8 filenames but same as `ls`. I think it’s ready to merge.

**Review by cafkafk — APPROVED** [review]
My local testing of this and integration tests indicates this is working, ty for the pull request!

**Pull request merged**